### ISSUE TITLE
Ambiguous column error for remote() tables

### ISSUE THREAD

**Issue by zhuyoucui0909** [ClickHouse/ClickHouse#11135]
**Describe the issue**

There is ambiguous column error on version 20.3.8.53 for distributed table. The same sql run correct on version 19.14.7.15.

**How to reproduce**
To simplify the reproducing steps, I use remote query on two hosts to substitute the query on distributed table.

1. prepare table and data on host1 and host2

`create table test.fact (id1 Int64, id2 Int64, value Int64) ENGINE = MergeTree() ORDER BY id1;`
`create table test.dimension (id1 Int64, name String) ENGINE = MergeTree() ORDER BY id1;`
`insert into test.fact values (1,2,10),(2,2,10),(3,3,10),(4,3,10);`
`insert into test.dimension values (1,'name_1'),(2,'name_1'),(3,'name_3'),(4, 'name_4');`

2. run sql on both host1 and host2 is correct
`SELECT 
    f.id1 AS ID, 
    d.name AS Name, 
    sum(f.value)
FROM test.fact AS f
LEFT JOIN test.dimension AS d ON f.id1 = d.id1
WHERE f.id1 = f.id2
GROUP BY 
    ID, 
    Name`

Result

┌─ID─┬─Name───┬─sum(value)─┐
│  2 │ name_1 │         10 │
│  3 │ name_3 │         10 │
└────┴────────┴────────────┘

3. But, run the sql based on two remote hosts is not correct. Run on host1, reports error on host2. Run on host2, report error on host1.
`SELECT 
    f.id1 AS ID, 
    d.name AS Name, 
    sum(f.value)
FROM remote('{{host1}},{{host2}}', test.fact, 'default', '') AS f
LEFT JOIN test.dimension AS d ON f.id1 = d.id1
WHERE f.id1 = f.id2
GROUP BY 
    ID, 
    Name`

Report error

`↖ Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.) Received exception from server (version 20.3.8):
Code: 352. DB::Exception: Received from localhost:9000. DB::Exception: Received from {host2}:9000. DB::Exception: Ambiguous column 'id1'. 
`

* Which ClickHouse server versions are incompatible
The same sql run correct on version 19.14.7.15, but fail on version 20.3.8.53. On version 20.3.8.53, it reports  "DB::Exception: Ambiguous column***"

* Which interface to use, if matters

clickhouse-client

**Additional context**
We are planing to upgrade clickhouse version from 19.14.7.15. to 20.3.8.53. 
So, we do function regression for the new version 20.3.8.53. We unfortunately found that our widely used sql pattern, as described above, is not supported on the new version.

*This issue is quite emergent on our side. Looking forward to your reply. Thanks!*

**Comment by zhuyoucui0909** [comment]
Any findings or update for this?

**Comment by zhuyoucui0909** [comment]
Do you know which version introduce this issue? We are encountering a server hung issue on version 19.14.7.15. We can try to upgrade to other version that without this back_compatibility issue.

**Comment by 4ertus2** [comment]
```it reports "DB::Exception: Ambiguous column***"```
Could you print stack trace after this line? You have omitted most the significant part of error.

**Comment by 4ertus2** [comment]
The 'Ambiguous column' error added cause tou may have wrong result without it in some cases. Unfortunatelly we didn't find that it breaks joins with remote() tables. I cannot find quiqly which version is first affected but the code in development branch committed at least 29.07.2019. So newer versions are probably affected.

**Comment by winter7** [comment]
@4ertus2 I reproduced this and got
```
exception:            Code: 352, e.displayText() = DB::Exception: Received from chi-ebstest-clickhouse-s0-r0-0.ebstest-clickhouse-shard0-0.clickhouse.svc.cluster.local:9000. DB::Exception: Ambiguous column 'id1'. Stack trace:

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x10542450 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x8f4272d in /usr/bin/clickhouse
2. ? @ 0xd0a837a in /usr/bin/clickhouse
3. ? @ 0xd29dfc7 in /usr/bin/clickhouse
4. DB::CrossToInnerJoinMatcher::visit(DB::ASTSelectQuery&, std::__1::shared_ptr<DB::IAST>&, DB::CrossToInnerJoinMatcher::Data&) @ 0xd29efbb in /usr/bin/clickhouse
5. DB::CrossToInnerJoinMatcher::visit(std::__1::shared_ptr<DB::IAST>&, DB::CrossToInnerJoinMatcher::Data&) @ 0xd29feb3 in /usr/bin/clickhouse
6. DB::InDepthNodeVisitor<DB::CrossToInnerJoinMatcher, true, std::__1::shared_ptr<DB::IAST> >::visit(std::__1::shared_ptr<DB::IAST>&) @ 0xd157d7b in /usr/bin/clickhouse
7. DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, DB::Context const&, std::__1::shared_ptr<DB::IBlockInputStream> const&, std::__1::optional<DB::Pipe>, std::__1::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0xd149fcd in /usr/bin/clickhouse
8. DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, DB::Context const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0xd14ab89 in /usr/bin/clickhouse
9. DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::__1::shared_ptr<DB::IAST> const&, DB::Context const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0xd34fcd6 in /usr/bin/clickhouse
10. DB::InterpreterFactory::get(std::__1::shared_ptr<DB::IAST>&, DB::Context&, DB::QueryProcessingStage::Enum) @ 0xd09efa4 in /usr/bin/clickhouse
11. ? @ 0xd55ee85 in /usr/bin/clickhouse
12. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, bool) @ 0xd561c71 in /usr/bin/clickhouse
13. DB::TCPHandler::runImpl() @ 0x9025489 in /usr/bin/clickhouse
14. DB::TCPHandler::run() @ 0x9026470 in /usr/bin/clickhouse
15. Poco::Net::TCPServerConnection::start() @ 0xe3b9a4b in /usr/bin/clickhouse
16. Poco::Net::TCPServerDispatcher::run() @ 0xe3b9ecd in /usr/bin/clickhouse
17. Poco::PooledThread::run() @ 0x105d05b7 in /usr/bin/clickhouse
18. Poco::ThreadImpl::runnableEntry(void*) @ 0x105cc3bc in /usr/bin/clickhouse
19. ? @ 0x105cdd5d in /usr/bin/clickhouse
20. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
21. __clone @ 0x12188f in /lib/x86_64-linux-gnu/libc-2.27.so
: While executing SourceFromInputStream (version 20.3.8.53 (official build))
stack_trace:          0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x10542450 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x8f4272d in /usr/bin/clickhouse
2. DB::readException(DB::ReadBuffer&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x8f78170 in /usr/bin/clickhouse
3. DB::Connection::receiveException() @ 0xdb365ef in /usr/bin/clickhouse
4. DB::Connection::receivePacket() @ 0xdb3b195 in /usr/bin/clickhouse
5. DB::MultiplexedConnections::receivePacket() @ 0xdb4ffea in /usr/bin/clickhouse
6. DB::RemoteBlockInputStream::readImpl() @ 0xce8bdf4 in /usr/bin/clickhouse
7. DB::IBlockInputStream::read() @ 0xce5701f in /usr/bin/clickhouse
8. DB::SourceFromInputStream::generate() @ 0xdefb3a4 in /usr/bin/clickhouse
9. DB::ISource::work() @ 0xdba8ceb in /usr/bin/clickhouse
10. DB::SourceFromInputStream::work() @ 0xdefb285 in /usr/bin/clickhouse
11. ? @ 0xdbe18c1 in /usr/bin/clickhouse
12. DB::PipelineExecutor::executeSingleThread(unsigned long, unsigned long) @ 0xdbe5a4d in /usr/bin/clickhouse
13. DB::PipelineExecutor::executeImpl(unsigned long) @ 0xdbe7af8 in /usr/bin/clickhouse
14. DB::PipelineExecutor::execute(unsigned long) @ 0xdbe7cc5 in /usr/bin/clickhouse
15. ? @ 0x901c207 in /usr/bin/clickhouse
16. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0x8f6792b in /usr/bin/clickhouse
17. ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()::operator()() const @ 0x8f68608 in /usr/bin/clickhouse
18. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x8f667eb in /usr/bin/clickhouse
19. ? @ 0x8f64c33 in /usr/bin/clickhouse
20. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
21. __clone @ 0x12188f in /lib/x86_64-linux-gnu/libc-2.27.so
```

**Comment by winter7** [comment]
I wonder if this really is a bug, and when do you expect to fix it?

**Comment by zhuyoucui0909** [comment]
Can you tell us what's the earliest time to release the fix? It is a stoping issue for us. Please tell us the schedule so we can arrange our job.
Best Regards.

**Comment by zhuyoucui0909** [comment]
We did a simple test on latest stable version v20.4.4.18. On this version, the issue is not reproduced. We will test more on the new version.

**Comment by 4ertus2** [comment]
Could you print it here? We could add it in our tests to avoid regression.

**Comment by zhuyoucui0909** [comment]
v20.4.4.18 is a good version for us. We have adopted to  v20.4.4.18 version.

You can add the reproduce steps on this issue to your regression. 
And you can also use distributed table to replace the remote table call.

**Comment by alexey-milovidov** [comment]
Ok. I have successfully reproduced this issue on 20.3.8 and checked that it is fixed in 20.3.12 and more recent versions.
